lib/cn/cn_tree_cursor.c

```diff
@@ -100,6 +100,9 @@ cn_tree_kvset_refs(struct cn_tree_node *node, struct cn_level_cursor *lcur)
     struct table *tab = lcur->cnlc_kvref_tab;
     struct kvset_list_entry *le;
 
+    if (!atomic_read(&node->tn_readers))
+        atomic_inc(&node->tn_readers);
```

Comment:
This use of atomics worries me.  What is the acronym?  TOCTOU?

Reply:
Now that I see its purpose, it's no longer concerning.

Reply:
Time of creation, time of use. Had the same initial thought

Reply:
Can we use `atomic_cmpxchg()` instead where old is 0 and new is 1?

Reply:
Right, there is no race.  The idea is to change the flag only if it's not set, so it's a read-only barrier-less operation most of the time.  A CAS would introduce a full r/w barrier which is not necessary for this purpose, and since it uses the lock prefix I suspect it would also require the current CPU to acquire exclusive access to the cache line, which would not be good.  I don't know for certain on that last point, would have to investigate...